python_modules/dagster/dagster/_core/storage/state/run_storage_state_storage.py

```diff
@@ -0,0 +1,43 @@
+from pathlib import Path
+from typing import TYPE_CHECKING, Optional
+
+from dagster_shared.serdes.serdes import deserialize_value, serialize_value
+
+from dagster._core.storage.state.base import StateStorage
+from dagster._core.storage.state.state_versions import StateVersions
+
+if TYPE_CHECKING:
+    from dagster._core.storage.runs.base import RunStorage
+
+
+class RunStorageStateStorage(StateStorage):
+    """Implements StateStorage using a RunStorage as the backing store."""
+
+    VERSIONS_KEY = "__latest_state_versions__"
+
+    def __init__(self, run_storage: "RunStorage"):
+        self._run_storage = run_storage
+
+    def get_latest_state_versions(self) -> Optional[StateVersions]:
+        raw_value = self._run_storage.get_cursor_values({self.VERSIONS_KEY}).get(self.VERSIONS_KEY)
+        if raw_value is None:
+            return None
+        return deserialize_value(raw_value, StateVersions)
+
+    def download_state_to_path(self, key: str, version: str, path: Path) -> None:
+        state_key = self._get_state_key(key, version)
+        state_value = self._run_storage.get_cursor_values({state_key}).get(state_key)
+        if state_value:
+            path.write_bytes(state_value.encode("utf-8"))
+            return
+        raise ValueError(f"No state found for key {key} and version {version}")
+
+    def set_latest_version(self, key: str, version: str) -> None:
+        self._run_storage.set_cursor_values(
+            {self.VERSIONS_KEY: serialize_value(self.get_updated_state_versions(key, version))}
+        )
+
+    def upload_state_from_path(self, key: str, version: str, path: Path) -> None:
+        state_key = self._get_state_key(key, version)
+        self._run_storage.set_cursor_values({state_key: path.read_bytes().decode("utf-8")})
```

Comment:
I know this is basically at parity with how we currently abuse the DB to store this sort of state, but I do think its worth thinking through how much work it actually would be to instead target a simple blob store backed version leveraging `universal-pathlib`.

I think it might just be taking the path prefix (aka bucket) as a bit of config and then making sure the errors are reasonable if you dont have the right `fsspec` extras installed. 

Assuming this state backed defs stuff is as well setup as it seems and we are able to leverage it for more and more I worry that this DB backed variant will haunt us in the future.